internal/controllers/machinedeployment/machinedeployment_rollout_sequence_test.go

```diff
@@ -0,0 +1,876 @@
+			return i
+		}
+	}
+}
+
+type fileLogger struct {
+	t *testing.T
+
+	testCase              string
+	fileName              string
+	testCaseStringBuilder strings.Builder
+}
+
+func newFileLogger(t *testing.T) *fileLogger {
+	t.Helper()
+	return &fileLogger{t: t, testCaseStringBuilder: strings.Builder{}}
+}
+
+func (l *fileLogger) NewTestCase(name, fileName string) {
+	if l.testCase != "" {
+		l.testCaseStringBuilder.Reset()
+	}
+	l.testCaseStringBuilder.WriteString(fmt.Sprintf("## %s\n\n", name))
+	l.testCase = name
+	l.fileName = fileName
+}
+
+func (l *fileLogger) Logf(format string, args ...interface{}) {
+	l.t.Logf(format, args...)
+
+	s := strings.TrimSuffix(fmt.Sprintf(format, args...), "\n")
+	sb := &strings.Builder{}
+	if strings.Contains(s, "\n") {
+		lines := strings.Split(s, "\n")
+		for _, line := range lines {
+			indent := "  "
+			if strings.HasPrefix(line, "[") {
+				indent = ""
+			}
+			sb.WriteString(indent + line + "\n")
+		}
+	} else {
+		sb.WriteString(s + "\n")
+	}
+	l.testCaseStringBuilder.WriteString(sb.String())
+}
+
+func (l *fileLogger) EndTestCase() (string, string, error) {
+	if err := os.WriteFile(fmt.Sprintf("%s.test.log", l.fileName), []byte(l.testCaseStringBuilder.String()), 0600); err != nil {
+		return "", "", err
+	}
+	if err := os.WriteFile(fmt.Sprintf("%s.test.log.golden", l.fileName), []byte(l.testCaseStringBuilder.String()), 0600); err != nil {
+		return "", "", err
+	}
+
+	currentBytes, _ := os.ReadFile(fmt.Sprintf("%s.test.log", l.fileName))
+	current := string(currentBytes)
+
+	goldenBytes, _ := os.ReadFile(fmt.Sprintf("%s.test.log.golden", l.fileName))
+	golden := string(goldenBytes)
+
+	return current, golden, nil
+}
+
+func sortMachineSetMachines(machines []*clusterv1.Machine) {
+	sort.Slice(machines, func(i, j int) bool {
+		iIndex, _ := strconv.Atoi(strings.TrimPrefix(machines[i].Name, "m"))
+		jiIndex, _ := strconv.Atoi(strings.TrimPrefix(machines[j].Name, "m"))
+		return iIndex < jiIndex
+	})
+}
+
+func minAvailableBreachToleration() func(log *fileLogger, _ int, _ *rolloutScope, _, _ int32) bool {
+	return func(log *fileLogger, _ int, _ *rolloutScope, _, _ int32) bool {
+		log.Logf("[Toleration] tolerate minAvailable breach")
+		return true
+	}
+}
+
+func maxSurgeToleration() func(log *fileLogger, _ int, _ *rolloutScope, _, _ int32) bool {
+	return func(log *fileLogger, _ int, _ *rolloutScope, _, _ int32) bool {
+		log.Logf("[Toleration] tolerate maxSurge breach")
+		return true
+	}
+}
+
+// default task order ensure the controllers are run in a consistent and predictable way: md, ms1, ms2 and so on.
+func defaultTaskOrder(current *rolloutScope) []int {
+	taskOrder := []int{}
+	for t := range len(current.machineSets) + 1 + 1 { // +1 is for the MachineSet that might be created when reconciling md, +1 is for the md itself
+		taskOrder = append(taskOrder, t)
+	}
+	return taskOrder
+}
+
+func randomTaskOrder(current *rolloutScope, rng *rand.Rand) []int {
+	u := &UniqueRand{
```

Comment:
Probably this can be further simplified when we inline the code from UniqueRand

Reply:
I think this will not improve readability.
UniqueRand takes care of UniqueRand generation; here we are solving different a different problem which is to have some non-determinism for skipping a task or repeating it (let's not mix this problems only because we are using the same randomizer)